rfc/rfc-53/rfc-53.md

```diff
@@ -0,0 +1,120 @@
+    using `BoundedInMemoryQueueConsumer` writing hudi data file.
+- [abstract] `HoodieExecutor`: Executor which orchestrates concurrent producers and consumers communicating through a inner message queue.
+The current implementations are as follows:
+    - [based-master] `BoundedInMemoryExecutor` takes as input the size limit, queue producer(s), consumer and transformer and exposes API to orchestrate concurrent execution of these actors communicating through a central LinkedBlockingQueue.
+    - [rfc-new] `DisruptorExecutor` Control the initialization, life cycle of the disruptor, and coordinate the work of the producer, consumer, and ringbuffer related to the disruptor, etc.
+    
+Secondly, This rfc implements disruptor related producers, message handlers and executor which use this lock-free message queue based on the above abstraction. Some compents are introduced in the first part. In this phase, we discuss how to use disruptor in hoodie writing stages.
+
+The Disruptor is a library that provides a concurrent ring buffer data structure. It is designed to provide a low-latency, high-throughput work queue in asynchronous event processing architectures.
+
+We use the Disruptor multi-producer single-consumer working model:
+- Define `DisruptorPublisher` to register producers into Disruptor and control the produce behaviors including life cycle.
+- Define `DisruptorMessageHandler` to register consumers into Disruptor and write consumption data from disruptor to hudi data file
+- Define `HoodieDisruptorEvent` as the carrier of the hoodie message
+- Define `HoodieDisruptorEventFactory`: Pre-populate all the hoodie events to fill the RingBuffer. 
+We can use `HoodieDisruptorEventFactory` to create `HoodieDisruptorEvent` storing the data for sharing during exchange or parallel coordination of an event.
+- Expose some necessary parameters for the users with a proper default to tune in different scenarios.
+
+Finally, let me introduce the new parameters:
+  - `hoodie.write.executor.type`: Choose the type of executor to use, which orchestrates concurrent producers and consumers communicating through a inner message queue. 
+  Default value is `BOUNDED_IN_MEMORY_EXECUTOR` which used a bounded in-memory queue `LinkedBlockingQueue`. 
+  Also users could use `DISRUPTOR_EXECUTOR`, which use disruptor as a lock free message queue to gain better writing performance. 
+  Although `DISRUPTOR_EXECUTOR` is still an experimental feature.
+  - `hoodie.write.buffer.size`: The size of the Disruptor Executor ring buffer, must be power of 2.
+  - `hoodie.write.wait.strategy`: Strategy employed for making DisruptorExecutor wait for a cursor.
+
+4. limitation
+  For now, this disruptor executor is only supported for spark insert and spark bulk insert operation. Other operations like spark upsert is still on going.
```

Comment:
other engine such as flink is still on going?

Reply:
Yes. Flink related writing is still on going as this feature is still in experimental.
But it will not take much effort to migrate when spark ingestion is ready :)

Reply:
AFAIK, flink does not use spark's producer/consumer model to write the data. So I am wondering how the plan rollout for Flink engine. 

By the way, have you tried to remove the producer/consumer at all? The writing actually is a blocking single producer - single consumer cases. I guess it could perform better in some cases, as we save all the overhead of the message queue.

Reply:
Hi @YuweiXiao Thanks a lot for your attention.
It seems that flink use something like `FlinkLazyInsertIterable` which hold `BoundedInMemoryExecutor` to do ingestion works same as spark.
So that maybe we can provide this new disruptorExecutor as a new option.

<img width="1667" alt="截屏2022-05-20 上午10 52 03" src="https://user-images.githubusercontent.com/69956021/169439622-a2143e44-b609-4daf-a4a2-f64741db258c.png">

Reply:
```
By the way, have you tried to remove the producer/consumer at all? The writing actually is a blocking single producer - single consumer cases. I guess it could perform better in some cases, as we save all the overhead of the message queue.
```
Nice Catch! It may happen when using bulk_insert as single producer and single consumer.
Yes, I can have a try as another option

Reply:
> By the way, have you tried to remove the producer/consumer at all? The writing actually is a blocking single producer - single consumer cases. I guess it could perform better in some cases, as we save all the overhead of the message queue.

Nice Catch! It may happen when using bulk_insert as single producer and single consumer.
Yes, I can have a try as another option

Reply:
Oh ok, didn't realize `FlinkLazyInsertIterable`. But looking at its usage, I found the removal of producer/consumer may be suitable for Flink engine. The producer is simply a iterator of `List<HoodieRecord>`, and the consumption is single-thread too.

Reply:
> > By the way, have you tried to remove the producer/consumer at all? The writing actually is a blocking single producer - single consumer cases. I guess it could perform better in some cases, as we save all the overhead of the message queue.
> 
> Nice Catch! It may happen when using bulk_insert as single producer and single consumer. Yes, I can have a try as another option

Can we also reflect it on the RFC ?

Reply:
Sure will reflect these idea in this RFC ASAP.
Thanks